gateway/hail.nginx.conf.in

```diff
@@ -158,3 +161,118 @@ server {
     include /etc/letsencrypt/options-ssl-nginx.conf;
     ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
 }
+
+server {
+    server_name app.@domain@;
+
+    gzip on;
+    gzip_types  text/plain text/css text/html text/javascript application/javascript application/x-javascript text/js text/xml  application/xml application/json application/rss+xml image/svg+xml font/opentype;
+
+    location / {
+        proxy_pass http://web/;
+    }
+
+    #listen 80;
+
+    listen [::]:443 ssl;
+    listen 443 ssl;
+    ssl_certificate /etc/letsencrypt/fullchain.pem;
+    ssl_certificate_key /etc/letsencrypt/privkey.pem;
+    include /etc/letsencrypt/options-ssl-nginx.conf;
+    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
+}
+
+server {
+    server_name notebook-api.@domain@;
+
+    # Support arge URI's stemming from access tokens
+    client_header_buffer_size 64k;
+    large_client_header_buffers 4 64k;
+
+    gzip on;
+    gzip_types  text/plain application/json;
+
+    location = /auth {
+        internal;
+
+        resolver kube-dns.kube-system.svc.cluster.local;
+        proxy_pass http://auth-gateway.default.svc.cluster.local/verify/$auth_request_uri;
+    }
+
+    location = /auth-notebook {
+        internal;
+
+        resolver kube-dns.kube-system.svc.cluster.local;
+        proxy_pass http://notebook-api.default.svc.cluster.local/api/verify/$svc_name/$auth_request_uri;
+    }
+
+    location / {
+        if ($request_method = 'OPTIONS') {
+            add_header 'Access-Control-Allow-Origin' '*';
+            add_header 'Access-Control-Allow-Credentials' 'true';
+            add_header 'Access-Control-Allow-Methods' 'GET, POST, DELETE, OPTIONS';
+            add_header 'Access-Control-Allow-Headers' 'Authorization,Keep-Alive,User-Agent,X-Requested-With,Cache-Control,Content-Type';
+            add_header 'Access-Control-Max-Age' 86400;
+            add_header 'Content-Type' 'text/plain charset=UTF-8';
+            add_header 'Content-Length' 0;
+            return 204;
+            break;
+        }
+
+        set $auth_request_uri "$is_args$args";
+        auth_request /auth;
+        auth_request_set $auth_user $upstream_http_user;
+        auth_request_set $auth_scope $upstream_http_scope;
+        proxy_pass http://notebook-api;
+        proxy_set_header Host $http_host;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header X-Forwarded-Host $http_host;
+        proxy_set_header X-Forwarded-Proto $scheme;
+        proxy_set_header X-Real-IP $remote_addr;
+        proxy_set_header Upgrade $http_upgrade;
+        proxy_set_header Connection "upgrade";
+        proxy_set_header User $auth_user;
+        proxy_set_header Scope $auth_scope;
+
+        proxy_http_version 1.1;
+        proxy_read_timeout 86400;
+        proxy_buffering off;
+    }
+
+    location ~ /instance/([^/]+)/(.*) {
+        set $svc_name $1;
+        #set $auth_request_uri "$is_args$args";
+        #auth_request /auth-notebook;
```

Comment:
Remove.

Reply:
> The auth_request for notebook redirect will be reenabled when I modify it to properly handle not only the initial redirect

Ah, I see.  Leaving this out means, essentially, anyone with the right URL can get execution access to our cluster.  This seems bad, even temporarily.  What's involved in fixing the endpoint to support the other requests?

Reply:
Right, yes, not safe. For an initial demo I don't think it's very problematic, since attackers need to guess a svc_name and a long randomly generated string, for resources that live only the length of the demo, and whose names are transmitted over https and not stored browser-side (in-memory only). Potential strategies:

1) in notebook-api/verify response, set a cookie containing the access_token
2) allow all .css/.js files, and for all other requests use the Referrer, which contains the access_token.
  * The .js files should contain no state, they're just application/ui libraries. Even if they were resources that are only served to authenticated users, they will not be compiled for each user specifically (most likely).
  * The requests with referrer headers are Jupyter-issued redirects to protected resources it seems, and those possess the access_token, and we can/will check them in this mode

We could also save a Flask-issued session cookie, encrypted with a secret, and check that to avoid re-validating the access_token.